homeassistant/config_entries.py

```diff
@@ -1169,6 +1174,45 @@ def _data_to_save(self) -> dict[str, list[dict[str, Any]]]:
         """Return data to save."""
         return {"entries": [entry.as_dict() for entry in self._entries.values()]}
 
+    async def async_remove_device(self, entry_id: str, device_id: str) -> bool:
+        """Remove a device."""
+        if (config_entry := self.async_get_entry(entry_id)) is None:
+            raise UnknownEntry
+
+        if not config_entry.supports_remove_device:
+            raise HomeAssistantError("Config entry does not support device removal")
+
+        dev_reg = device_registry.async_get(self.hass)
+
+        if (device_entry := dev_reg.async_get(device_id)) is None:
+            raise HomeAssistantError("Unknown device")
+
+        if entry_id not in device_entry.config_entries:
+            raise HomeAssistantError("Config entry not in device")
+
+        try:
+            integration = await loader.async_get_integration(
+                self.hass, config_entry.domain
+            )
+        except loader.IntegrationNotFound as exc:
+            raise HomeAssistantError("Integration not found") from exc
+        try:
+            component = integration.get_component()
+        except ImportError as exc:
+            raise HomeAssistantError("Component not found") from exc
+
+        # Check with the config entry if it can be removed from the device
+        if not await component.async_remove_config_entry_device(
```

Comment:
I really hope we can do it without this function, like checking if there are no entities or they are all restored

Reply:
That's not possible since not all integrations know when a device should be removed.

Eg the mysensors integration keeps track of all presented sensors and has custom restore to avoid the user having to present the sensors again after each Home Assistant restart. There's no way of asking battery powered sensors if they are still in the network. So it's not possible to automatically detect if a sensor has been removed from the network or if it's just sleeping.

Users want to be able to remove these kind of devices manually as they know what the truth is.

Reply:
It's the same problem with cast: the device or group might be offline for an extended period or the device or group is actually removed and we don't know which case it is but the user does.

Reply:
Alright. In that case let's just have the frontend be able to query this value for a config entry/device combo. No need to guard the call itself. That way we can conditionally show the "Remove device" on the device info page.

Let's also call it more descriptive that it is a check if it's ok, maybe `async_remove_config_entry_device_allowed`?

Reply:
I was too focused on cast, which indeed does not need to take any actions when it's removed from a device.
In case of for example MQTT and Tasmota, the integration will need to know it's being removed so it can clear the discovery topic from the broker or the device will be rediscovered.

Reply:
Dito for mysensors.

Reply:
We could also have the integrations listen to `EVENT_DEVICE_REGISTRY_UPDATED`, but we would then need to update that event so it's possible to see what has changed just like we do for `EVENT_ENTITY_REGISTRY_UPDATED`.

Reply:
The downside with that is that integrations would act after the fact that the device was removed in the registry instead of before. That may not be a requirement for this feature but it's a change in feature.

Reply:
Yes, that should work. :+1:

Reply:
Just to confirm, do we change the entire PR, so we add a WS API `device_registry/remove_config_entry` instead of `config_entries/remove_device`? Or do we keep the PR as is, but just move the check for `component.async_remove_config_entry_device` to `DeviceRegistry.async_update_device`?

Reply:
I would move it all in the device registry files

Reply:
OK, refactored accordingly in 65159d6